crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs

```diff
@@ -0,0 +1,3454 @@
+                || is_expression_referentially_unique(&logical.right)
+        }
+        Expression::BinaryExpression(bin_expr) => {
+            is_expression_referentially_unique(&bin_expr.right)
+        }
+        Expression::AssignmentExpression(assignment) => {
+            is_expression_referentially_unique(&assignment.right)
+        }
+        _ => false,
+    }
+}
+
+#[derive(Debug)]
+enum CallbackNode<'a> {
+    Function(&'a Function<'a>),
+    ArrowFunction(&'a ArrowFunctionExpression<'a>),
+}
+
+impl<'a> CallbackNode<'a> {
+    fn is_async(&self) -> bool {
+        match self {
+            CallbackNode::Function(func) => func.r#async,
+            CallbackNode::ArrowFunction(func) => func.r#async,
+        }
+    }
+
+    fn body(&self) -> Option<&FunctionBody<'a>> {
+        match self {
+            CallbackNode::Function(func) => func.body.as_deref(),
+            CallbackNode::ArrowFunction(func) => Some(&func.body),
+        }
+    }
+}
+
+impl<'a> GetSpan for CallbackNode<'a> {
+    fn span(&self) -> Span {
+        match self {
+            CallbackNode::Function(func) => func.span,
+            CallbackNode::ArrowFunction(func) => func.span,
+        }
+    }
+}
+
+// https://github.com/facebook/react/blob/1b0132c05acabae5aebd32c2cadddfb16bda70bc/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L1789
+fn get_reactive_hook_callback_index(node: &CallExpression) -> Option<usize> {
+    let node_name = get_node_name_without_react_namespace(&node.callee);
+
+    let hook_name = node_name?;
+
+    match hook_name.as_str() {
+        "useEffect" | "useLayoutEffect" | "useCallback" | "useMemo" => Some(0),
+        "useImperativeHandle" => Some(1),
+        _ => {
+            // TODO: custom nodes
+            None
+        }
+    }
+}
+
+fn get_node_name_without_react_namespace<'a, 'b>(expr: &'b Expression<'a>) -> Option<&'b Atom<'a>> {
+    match expr {
+        Expression::StaticMemberExpression(member) => {
+            if let Expression::Identifier(_ident) = &member.object {
+                return Some(&member.property.name);
+            }
+            None
+        }
+        Expression::Identifier(ident) => Some(&ident.name),
+        _ => None,
+    }
+}
+
+#[derive(Debug)]
+struct Dependency<'a> {
+    iref: &'a IdentifierReference<'a>,
+    // the symbol id that this dependency is referring to
+    symbol_id: Option<SymbolId>,
+    chain: Vec<Atom<'a>>,
+}
+
+impl Hash for Dependency<'_> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.iref.name.hash(state);
+        self.symbol_id.hash(state);
+        self.chain.hash(state);
+    }
+}
+
+impl PartialEq for Dependency<'_> {
+    fn eq(&self, other: &Self) -> bool {
+        self.iref.name == other.iref.name && self.chain == other.chain
+    }
+}
+
+impl Eq for Dependency<'_> {}
+
+impl Dependency<'_> {
+    fn to_string(&self) -> String {
+        [vec![self.iref.name.clone()], self.chain.clone()].concat().join(".")
```

Comment:
I would recommend avoiding creating a vec here, can we push to a string directly instead using `String.push` and `String.push_str`?

Reply:
updated to use an iterator
```rs
std::iter::once(&self.iref.name)
    .chain(self.chain.iter())
    .map(|s| s.as_str())
    .join(".")
```

which should avoid the vec?